opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/__init__.py

```diff
@@ -0,0 +1,93 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from threading import Lock
+
+
+class Mapping(ABC):
+
+    # pylint: disable=no-member
+    def __new__(cls, scale: int):
+
+        if not hasattr(cls, "_mappings"):
+            cls._mappings = {}
+
+        if not hasattr(cls, "_mappings_lock"):
+            cls._mappings_lock = Lock()
+
+        with cls._mappings_lock:
```

Comment:
There is a race here to create the initial `cls._mappings` and `cls._mappings_lock`.

My two cents: it's easier to understand a factory providing the correct mapping impl given the scale parameter and handling the memoization of instances. Especially looking at the child instances, it's hard for a reader to tell that `ExponentMapping(scale)` is returning a memoized instance.

Also `__new__()` behavior is not very intuitive, for example `__init__()` is still called every time regardless of the check on L32:

```python
class Foo:
    def __new__(cls, *args, **kwargs):
        if hasattr(cls, "_instance"):
            print("reutrning cached")
            return cls._instance
        print("creating new instance")
        inst = super().__new__(cls, *args, **kwargs)
        cls._instance = inst
        return inst

    def __init__(self):
        print("__init__() called")


Foo() # created for the first time
Foo() # return cached instance, but __init__() is still called
```

```console
creating new instance
__init__() called
reutrning cached
__init__() called
```

is that intentional?

Reply:
I have reverted the commit that moved the instantiation of the lock to the parent class, and just added a comment explaining where that lock is. That should fix the race.

We want to use a singleton here, that is why we use `__new__`. We intentionally make the class return always the same instance when instantiated, which is what the pattern intends. It is not obvious for the user that the child class returns a memoized instance, and that is the idea, it looks like it is not from the user perspective, but we optimize our code by don't instantiating new objects when not necessary. Hiding that from the user is precisely what we intend.

A factory function would be named something like `get_same_instance_of_logarigthm_mapping` which defeats the purpose of hiding the implementation details from the user and forces the user to know a new way of instantiating the class.

As for `__init__`, in this particular case there is no consequence to call `__init__` multiple times but I agree is nicer not to do so, refactored the code to only initialize these objects once, so I have refactored the initialization process.

Reply:
This singleton impl has:
- `__new__()`
- `_init()`
- `__init__()`
- `_mappings` and `_mappings_lock` in each subclass
- `__init__()` is still called every time the subclass is instantiated

Imo that is not easy to understand, would like to hear other folks thoughts. The cache factory would be very simple:

```py
@lru_cache
def _create_mapping(scale: int) -> Mapping:
  if scale < 0:
    return ExponentMapping(scale)
  else:
    return LogarithmMapping(scale)
```

(although IIRC `lru_cache` is not threadsafe which complicates things)

Reply:
I have removed the over and underflow errors from the API, everything related to exponential histogram now is in `_internal` and can be changed later. This is an implementation detail and since this PR has enough approvals I would prefer to merge.

Please open a new issue if you want changes here :slightly_smiling_face: